### PR TITLE
Desktop: Fix #5981: Scroll positions are not preserved when layout changes in v2.6.10

### LINKED ISSUE #5981: Desktop: [Regression][v2.6.10] Scroll positions are not preserved when layout changes
When the editor layout is changed, scroll positions became preserved in v2.6.9 by PR #5826, but they are not preserved in v2.6.10. It's a regression bug. 

I'm trying to fix it now.

Ref. 
- [Desktop: Scroll position is not remembered/preserved in Markdown Editor and Viewer · Issue #5708 · laurent22/joplin](https://github.com/laurent22/joplin/issues/5708)
- [Desktop: Fix #5708: Scroll positions are remembered/preserved (Line-Percent-based Sync Scroll) by ken1kob · Pull Request #5826 · laurent22/joplin](https://github.com/laurent22/joplin/pull/5826)

### PR DESCRIPTION
**Pull request by ken1kob**
This PR  fixes the regression bug #5981 happens in v2.6.10.

To preserve Editor's scroll positions when the editor layout changes, the following are implemented:
- When heights of lines in CodeMirror are changed, Editor's scroll positions are restored.
- Uncertain scroll positions are ignored in Editor.
- When Viewer is resized, its scroll position is sent to Editor. (The reverse is not necessary.)

The details are described in #5981.

### CODE REVIEW

File: packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts [hunk 1]
```diff
@@ -7,7 +7,10 @@ export default function useScrollHandler(editorRef: any, webviewRef: any, onScro
 	const ignoreNextEditorScrollTime_ = useRef(Date.now());
 	const ignoreNextEditorScrollEventCount_ = useRef(0);
 	const delayedSetEditorPercentScrollTimeoutID_ = useRef(null);
+	const scrollTopIsUncertain_ = useRef(true);
 	const lastResizeHeight_ = useRef(NaN);
+	const lastLinesHeight_ = useRef(NaN);
+	const restoreEditorPercentScrollTimeoutId_ = useRef<any>(null);
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: I'm a bit concerned about having so many refs in this hook. In general I feel that a ref has to be the absolute last desperate attempt to get thing working when nothing else does, but is that the case for this hook?

With everything being mutable it makes it hard to follow the flow since any variable can be modified anywhere, and the dependent hooks may or may not be called at the right time, or it can be random.

So I'm wondering if it's possible to try to get these 3 new variables as regular state variables?

**Review by laurent22 — COMMENTED** [review]
Thanks @ken1kob, I've just left one comment.

File: packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts (hunk 1 above)
Comment (ken1kob) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: > In general I feel that a ref has to be the absolute last desperate attempt to get thing working when nothing else does, but is that the case for this hook?

For pure declarative function components, useRef should be avoided. Otherwise, there are little reasons to avoid useRef.

In this case, to absorb CodeMirror's design issues, it is inevitable to have a lot of state variables. So the function component is essentially not declarative but imperative. 

> With everything being mutable it makes it hard to follow the flow since any variable can be modified anywhere, and the dependent hooks may or may not be called at the right time, or it can be random.

Since these useRef variables are not exported, their scope is limited within `useScrollHandler` function. They correspond to private fields in a class. Data encapsulation is not broken, and redundant dependencies are not introduced.

From the performance viewpoint, using many useRef variables introduces little impact.

> So I'm wondering if it's possible to try to get these 3 new variables as regular state variables?

Do you mean the following rewriting? If you decide there is no possibility of having multiple components of Markdown Editor in the future, they can be rewritten.

FROM:
```js
function useFunc() {
  const a = useRef(0);
  const b = useRef(0);
	...
}
```

TO:
```js
let a = 0;
let b = 0;

function useFunc() {
  ...
}
```

Another choice will be to encapsulate states into a dedicated class and keeps useRef of the class. Anyway, there are little differences between them from the viewpoints of performance and readability.

Comment (laurent22) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: > Since these useRef variables are not exported, their scope is limited within useScrollHandler function.

The variables aren't exported, but handlers that depend on them, such as `editor_update` are, so we have no guarantee that wherever it's called, it's not using stale data. So this use of refs in fact can leak to the entire application.

Using state variables (as opposed to refs) give strong guarantees that everything is up to date in the application, while with refs we need to track their changes and manually update components, or hope that they are going to be updated as a side-effect of something else (which of course can easily break).

> Do you mean the following rewriting?

No I mean to change `userRef` to `useState`, so for example:

`const [lastLinesHeight, setLastLineHeight] = useState(NaN);`

I think at least for some of your new variables you can use the state.

Comment (ken1kob) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: > so we have no guarantee that wherever it's called, it's not using stale data. So this use of refs in fact can leak to the entire application.

This senario may be that useRef variables are inappropriately used where useState variables are appropriate. Usecases of useRef is different from those of useState.

Typical usecases of useRef variables are:
- Simple data containers for a function component (like members of a class)
- Non-reactive data containers, which don't trigger any rendering when changed. They are useful to reduce redundant rendering.

> I think at least for some of your new variables you can use the state.

My answer is no. They cannot be replaced by useState. Their changes must not trigger rendering. They are all passive and used through event handlers activated by CodeMirror.

Comment (laurent22) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: You may be misunderstanding a bit how state variables works. They aren't going to trigger any rendering if they are not used by any hook or React component - and if they are, the component should indeed be updated.

I'd like to suggest that you try to change to useState and most likely you won't notice any unnecessary rendering. You can use something like useEffectDebugger or add console statements to verify. It would however be future proof because any future dependency of this hook (directly or indirectly) will be correctly updated when something changes.

Comment (ken1kob) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: > They aren't going to trigger any rendering if they are not used by any hook or React component

It sounds strange that useState variables "are not used by any hook or React component". Correctly declared useState variables are always coupled with the React component where its `useState()` is called directly or indirectly through a custom hook. For example, the above `lastLinesHeight` useState variable will be coupled with `CodeMirror` function component through `useScrollHander` custom hook. Whenever it is modified, `CodeMirror` and its child `Editor` will be re-rendered.

> I'd like to suggest that you try to change to useState

It was already considered. As mentioned above, replacing useRef to useState introduced unnecessary rendering of `CodeMirror` and `Editor`.

Additionally speaking, I found some unnecessary rendering in whole Joplin, and some of high-impact ones are removed in PR #5770. Of course, I won't say all reduandant rendering should be removed. Only high-impact ones deserve paying attentions. Some of the others should be left for maintenability.

>  It would however be future proof because any future dependency of this hook (directly or indirectly) will be correctly updated when something changes.

I agree. Future proofing is a good practice. In the case that useRef and useState are exchangeable, taking useState is almost better.

Comment (laurent22) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts:13]: Thanks for looking into it. I can't help but think there's something off with a hook that's entirely based on refs. It's probably partly due to scroll being special, or maybe it shouldn't be a hook but a simple utility function. Anyway, I think it's fine otherwise so let's merge.
